### ISSUE TITLE
Nightmare Oratuss logs failing to parse (Sequence contains no elements)

### ISSUE THREAD

**Issue by Sejsel** [baaron4/GW2-Elite-Insights-Parser#503]
Found some Nightmare Oratuss logs that result in an exception.
Found when comparing results with https://github.com/gw2scratch/evtc/.

```
Exception: System.InvalidOperationException: Sequence contains no elements
   at System.Linq.ThrowHelper.ThrowNoElementsException()
   at System.Linq.Enumerable.Min[TSource,TResult](IEnumerable`1 source, Func`2 selector)
   at GW2EIEvtcParser.EvtcParser.CompletePlayers() in /home/sejsel/programming/projects/current/gw2/ScratchEVTCParser/EI/GW2EIEvtcParser/EvtcParser.cs:line 672
   at GW2EIEvtcParser.EvtcParser.CompleteAgents() in /home/sejsel/programming/projects/current/gw2/ScratchEVTCParser/EI/GW2EIEvtcParser/EvtcParser.cs:line 710
   at GW2EIEvtcParser.EvtcParser.ParseLog(ParserController operation, Stream stream) in /home/sejsel/programming/projects/current/gw2/ScratchEVTCParser/EI/GW2EIEvtcParser/EvtcParser.cs:line 103
   at GW2EIEvtcParser.EvtcParser.ParseLog(ParserController operation, FileInfo evtc) in /home/sejsel/programming/projects/current/gw2/ScratchEVTCParser/EI/GW2EIEvtcParser/EvtcParser.cs:line 78
```

[20181017-005436.evtc.zip](https://github.com/baaron4/GW2-Elite-Insights-Parser/files/5632317/20181017-005436.evtc.zip)
[20181017-005754.evtc.zip](https://github.com/baaron4/GW2-Elite-Insights-Parser/files/5632318/20181017-005754.evtc.zip)
[20181017-010054.evtc.zip](https://github.com/baaron4/GW2-Elite-Insights-Parser/files/5632319/20181017-010054.evtc.zip)

**Comment by EliphasNUIT** [comment]
Interesting, there are no players in any of those. Looks like I'm doing the player presence check after CompletePlayers(), I'll move the block above it.

**Comment by Sejsel** [comment]
Also found a few SH logs with the same issue (exception thrown on the same line of code), but I guess your fix should solve all of these.

**Comment by EliphasNUIT** [comment]
Generally speaking, I need to overhaul the exception throwing in ParseLog. It was fine before when it was inside the main app that managed the parsing through tasks and checked IsFaulted on the task to display the problem or the operation canceled message on the UI. Right now, it's not very confortable for people using it in a standalone manner.

I intend to transform the signature of the method to
ParsedEvtcLog ParseLog(ParserController operation, FileInfo evtc, out ParsingFailureReason parsingFailureReason)

with the method returning null when parsing has failed. ParsingFailureReason will then contain utilities to :
- throw the Exception as it was done before
- throw the Exception if the reason is not an EIException (a bug or due to an agressive defensive programming, regardless an issue should be opened on such cases)
- Some attributes to get information as to what went wrong: the message, IsEvtcContentIssue, IsSafeToIgnore and IsParserBug

As such, the caller will decide what do to based on the ParsingFailureReason's content intead of having to catch random exception themself.

I'm also overhauling thrown exceptions by putting every custom exceptions under EIException and introducing EvtcContentException declined into:
- EvtcAgentException -> something is wrong with parsed agents
- EvtcCombatEventException -> something is wrong with parsed combat events
- EvtcFileException -> something is wrong with the file (not evtc, file does not exist, etc)
- MissingKeyActorsException -> species id mandatory for the encounter are not present in the agent pool

TooLong, TooShort and Skip will remain as such but has been regrouped under EINonFatalException for convenience.

**Comment by EliphasNUIT** [comment]
Fixed with #509
